csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs

```diff
@@ -75,6 +75,32 @@ public static SessionOptions MakeSessionOptionWithCudaProvider(int deviceId = 0)
             return options;
         }
 
+        /// <summary>
+        /// A helper method to construct a SessionOptions object for CUDA execution.
+        /// Use only if CUDA is installed and you have the onnxruntime package specific to this Execution Provider.
+        /// </summary>
+        /// <param name="cuda_options">CUDA EP provider options to configure the CUDA EP instance</param>>
+        /// <returns>A SessionsOptions() object configured for execution with CUDA provider options.</returns>
+        public static SessionOptions MakeSessionOptionWithCudaProvider(OrtCUDAProviderOptions cuda_options)
+        {
+            CheckCudaExecutionProviderDLLs();
+
+            OrtCUDAProviderOptionsNative cuda_options_native;
+            cuda_options_native.device_id = cuda_options.device_id;
+            cuda_options_native.cudnn_conv_algo_search = cuda_options.cudnn_conv_algo_search;
+            cuda_options_native.gpu_mem_limit = cuda_options.gpu_mem_limit;
+            cuda_options_native.arena_extend_strategy = cuda_options.arena_extend_strategy;
+            cuda_options_native.do_copy_in_default_stream = cuda_options.do_copy_in_default_stream;
+            cuda_options_native.has_user_compute_stream = 0;
+            cuda_options_native.user_compute_stream = IntPtr.Zero;
+            using (SessionOptions options = new SessionOptions())
+            {
+                NativeApiStatus.VerifySuccess(NativeMethods.SessionOptionsAppendExecutionProvider_CUDA(options.Handle, ref cuda_options_native));
+                NativeApiStatus.VerifySuccess(NativeMethods.OrtSessionOptionsAppendExecutionProvider_CPU(options.Handle, 1));
+                return options;
```

Comment:
>return options; [](start = 16, length = 15)

So using() clause would dispose of SessionOptions and you would return a disposed (dead) object, correct?

Reply:
You are right. I'm being cursory. I've modified this function to its original state.

To answer your previous question - What happens to this instance when either of the below native methods return an error VerifySuccess throws?
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.releasehandle?view=net-5.0
My understanding is that if VerifySuccess throws exception, still this instance will be handled by garbage collector. Due to this instance is SafeHandle, gc will call ReleaseHandle after normal finalizers have been run for objects that were garbage collected at the same time. Therefore, SessionOptions instance will be reclaimed. 

I'm also thinking about using following code:
```
try {
  SessionOptions options = new SessionOptions();
  NativeApiStatus.VerifySuccess(NativeMethods.OrtSessionOptionsAppendExecutionProvider_ROCM(options.Handle, deviceId));
  NativeApiStatus.VerifySuccess(NativeMethods.OrtSessionOptionsAppendExecutionProvider_CPU(options.Handle, 1));
}
catch (Exception e) {
  options.Dispose();
}
```
but I think it's not necessary. What do you think?

Reply:
This would be a correct imlpementation. You will need to add rethrow to the exception. You want to return options only when everything succeeds.

---
In reply to: [613934184](https://github.com/microsoft/onnxruntime/pull/7315#discussion_r613934184) [](ancestors = 613934184)